core/src/retransmit_stage.rs

```diff
@@ -209,49 +213,69 @@ pub type ShredFilter = LruCache<(Slot, u32, bool), Vec<u64>>;
 
 pub type ShredFilterAndHasher = (ShredFilter, PacketHasher);
 
-// Return true if shred is already received and should skip retransmit
+// Returns None if shred is already received and should skip retransmit.
+// Otherwise returns shred's slot.
 fn check_if_already_received(
     packet: &Packet,
-    shreds_received: &Arc<Mutex<ShredFilterAndHasher>>,
-) -> bool {
-    match get_shred_slot_index_type(packet, &mut ShredFetchStats::default()) {
-        Some(slot_index) => {
-            let mut received = shreds_received.lock().unwrap();
-            let hasher = received.1.clone();
-            if let Some(sent) = received.0.get_mut(&slot_index) {
-                if sent.len() < MAX_DUPLICATE_COUNT {
-                    let hash = hasher.hash_packet(packet);
-                    if sent.contains(&hash) {
-                        return true;
-                    }
-
-                    sent.push(hash);
-                } else {
-                    return true;
-                }
+    shreds_received: &Mutex<ShredFilterAndHasher>,
+) -> Option<Slot> {
+    let shred = get_shred_slot_index_type(packet, &mut ShredFetchStats::default())?;
+    let mut shreds_received = shreds_received.lock().unwrap();
+    let (cache, hasher) = shreds_received.deref_mut();
+    match cache.get_mut(&shred) {
+        Some(sent) if sent.len() >= MAX_DUPLICATE_COUNT => None,
+        Some(sent) => {
+            let hash = hasher.hash_packet(packet);
+            if sent.contains(&hash) {
+                None
             } else {
-                let hash = hasher.hash_packet(&packet);
-                received.0.put(slot_index, vec![hash]);
+                sent.push(hash);
+                Some(shred.0)
             }
+        }
+        None => {
+            let hash = hasher.hash_packet(packet);
+            cache.put(shred, vec![hash]);
+            Some(shred.0)
+        }
+    }
+}
 
-            false
+// Returns true if turbine retransmit peers patch (#14565) is enabled.
+fn enable_turbine_retransmit_peers_patch(bank_forks: &RwLock<BankForks>, shred_slot: Slot) -> bool {
+    let bank = {
+        let bank_forks = bank_forks.read().unwrap();
```

Comment:
instead of grabbing this `bank_forks` lock everytime, could we save the `feature_slot` and clone out the `epoch_schedule` to be used for `get_epoch()`

Reply:
the constraint is that we need to look up the bank for the `shred_slot`, where the `shred_slot` may change from one packet to the other.
I updated the code by adding a local cache for `slot -> bank` mapping to avoid locking `bank_forks` each time.

Reply:
got it, the bank needs to be continually updated to check for feature updates. I think we can do a little better than checking for a bank every slot though, since features are only updated at epoch boundaries: https://github.com/solana-labs/solana/blob/master/runtime/src/bank.rs#L1073.  

So If we clone out the `epoch_schedule` on startup: https://github.com/solana-labs/solana/pull/14631/files#diff-8d73ca7d46b0b2dba59642e0e183fbb6127c8459ade483df4ac729390519e6f6R266 which doesn't change from bank to bank, then we can detect when a `shred_slot` is in a new epoch and then only grab a new bank then.

Reply:
The concern with relying on:
* the `epoch_schedule` not changing from bank to bank.
* features getting activated only at epoch boundaries.

is that above two are some internal implementation details in another crate. They are not part of the api, and could easily change by someone unaware of the code here. So it opens door to bugs in the future.

This function locks `bank_forks` once per *unique* shred-slot but besides that it is already O(1). I feel like we at least need some evidence that this is going to be a bottleneck, before optimizing it further by relying on internals of another crate.